package/MDAnalysis/coordinates/H5MD.py

```diff
@@ -0,0 +1,587 @@
+        """converts units from H5MD to MDAnalysis notation
+        and fills units dictionary"""
+
+        # need this dictionary to associate 'position': 'length'
+        _group_unit_dict = {'time': 'time',
+                            'position': 'length',
+                            'velocity': 'velocity',
+                            'force': 'force'
+                            }
+
+        for group, unit in _group_unit_dict.items():
+            self._translate_h5md_units(group, unit)
+            self._check_units(group, unit)
+
+    def _check_units(self, group, unit):
+        """Raises error if no units are provided from H5MD file
+        and convert_units=True"""
+
+        if not self.convert_units:
+            return
+
+        errmsg = "H5MD file must have readable units if ``convert_units`` is"
+        " set to ``True``. MDAnalysis sets ``convert_units=True`` by default."
+        " Set ``convert_units=False`` to load Universe without units."
+
+        if unit == 'time':
+            if self.units['time'] is None:
+                raise ValueError(errmsg)
+
+        else:
+            if self._has[group]:
+                if self.units[unit] is None:
+                    raise ValueError(errmsg)
+
+    def _translate_h5md_units(self, group, unit):
+        """stores the translated unit string into the units dictionary"""
+
+        errmsg = "{} unit '{}' is not recognized by H5MDReader. Please raise"
+        " an issue in https://github.com/MDAnalysis/mdanalysis/issues"
+
+        # doing time unit separately because time has to fish for
+        # first available parent group - either position, velocity, or force
+        if unit == 'time':
+            for name, value in self._has.items():
+                if value:
+                    if 'units' in self._particle_group[name]['time'].attrs:
+                        try:
+                            self.units['time'] = self._unit_translation[
+                            'time'][self._particle_group[name][
+                            'time'].attrs['units']]
+                            break
+                        except KeyError:
+                            raise RuntimeError(errmsg.format(unit,
+                                               self._particle_group[
+                                               name]['time'].attrs[
+                                               'units'])) from None
+
+        else:
+            if self._has[group]:
+                if 'units' in self._particle_group[group].attrs:
+                    try:
+                        self.units[unit] = self._unit_translation[unit][
+                        self._particle_group[group].attrs['units']]
+                    except KeyError:
+                        raise RuntimeError(errmsg.format(unit,
+                                           self._particle_group[group].attrs[
+                                           'units'])) from None
+
+    def _copy_to_data(self):
```

Comment:
Move near `read_frame()` so that one doesn't have to look too far for the helper method.